stm32/Src/sdi12.c

```diff
@@ -3,54 +3,108 @@
+    while (HAL_LPTIM_ReadCounter(&hlptim1) < target_tick) {
+        // Wait
+    }
     
-    __HAL_RCC_GPIOA_CLK_ENABLE(); // Enable GPIOA clock
+}
+
+void TIM16_Delay_ms(uint32_t milliseconds) {
+    // Calculate the number of timer counts for specific pulse
+    
+    HAL_TIM_Base_Start(&htim16);
+    // Get the current value of the TIM16 counter
+    uint32_t start_tick = __HAL_TIM_GET_COUNTER(&htim16);
+
+    // Calculate the maximum target value of the TIM16 counter after the delay
+    uint32_t max_target_tick = 65535;
+
+    // Calculate the target value of the TIM16 counter after the delay
+    uint32_t target_tick = start_tick + milliseconds;
+
+    // Monitor the counter until the specified number of milliseconds have passed
+    while (__HAL_TIM_GET_COUNTER(&htim16) < target_tick) {
+        // Check for overflow
+        if (target_tick > max_target_tick) {
+            // Calculate the remaining counts after overflow
+            uint32_t remaining_counts = target_tick - max_target_tick;
+            // Wait until the remaining counts have also elapsed
+            while (__HAL_TIM_GET_COUNTER(&htim16) < remaining_counts) {
+                // Wait
+            }
+            // Exit the loop after handling overflow
+            break;
+        }
+    }
+    HAL_TIM_Base_Stop(&htim16);
+}
+
+void TIM1_Delay_ms(uint32_t milliseconds) {
+    // Calculate the number of timer counts for specific pulse
     
-    // Configure GPIO pin as output
-    GPIO_InitStruct.Pin = GPIO_PIN_2;
-    GPIO_InitStruct.Mode = GPIO_MODE_OUTPUT_PP;
-    GPIO_InitStruct.Speed = GPIO_SPEED_FREQ_LOW;
-    HAL_GPIO_Init(GPIOA, &GPIO_InitStruct);
+    HAL_TIM_OnePulse_Start(&htim1, TIM_CHANNEL_1); // Needs a channel parameter but it should be disabled according to CubeMX
+    // Get the current value of the TIM16 counter
+    uint32_t start_tick = __HAL_TIM_GET_COUNTER(&htim16);
+
+    // Calculate the maximum target value of the TIM16 counter after the delay
+    while(__HAL_TIM_GET_COUNTER(&htim1) <= __HAL_TIM_GET_COMPARE(&htim1, TIM_CHANNEL_1)){
+
+    }
 }
 
-// Function to configure GPIO as input
-void configureAsInput(void) {
-    GPIO_InitTypeDef GPIO_InitStruct = {0};
 
-    __HAL_RCC_GPIOA_CLK_ENABLE(); // Enable GPIOA clock
+void SDI12_WakeSensors(void){
+    HAL_LIN_SendBreak(&huart2); // Send a break 
+    HAL_GPIO_WritePin(GPIOC, GPIO_PIN_1, GPIO_PIN_RESET); // Send the marking 
+    HAL_Delay(20); // Need an extra 10ms to account for the fact that HAL_LIN_SendBreak is nonblocking
+}
 
-    // Configure GPIO pin as input
-    GPIO_InitStruct.Pin = GPIO_PIN_2;
-    GPIO_InitStruct.Mode = GPIO_MODE_INPUT;
-    GPIO_InitStruct.Pull = GPIO_NOPULL;
-    HAL_GPIO_Init(GPIOA, &GPIO_InitStruct);
+HAL_StatusTypeDef SDI12_SendCommand(const char *command, uint8_t size){
+    HAL_StatusTypeDef ret;
+    SDI12_WakeSensors();
+    ret = HAL_UART_Transmit(&huart2, (const uint8_t *) command, size, 1000);
```

Comment:
Make the timeout some static const so that its shared between all commands and can be configured easily